docs/express-vs-core.qmd

```diff
@@ -0,0 +1,262 @@
+
+##### Express advantages
+
+- No need to carefully match the names of placeholders and rendering functions.
+
+### Different syntax for UI containers
+
+Broadly speaking, there are two kinds of UI components in Shiny: _container components_, which, as the name suggests, can contain other components, and _non-container components_, which cannot.
+(You can also think of the UI as a tree data structure: container components have children, while non-container components are leaf nodes in the tree.)
+
+Here are some examples of container components:
+
+- [`ui.sidebar()`](../layouts/sidebars.qmd)
+- [`ui.card()`](../layouts/panels-cards.qmd#content-divided-by-cards)
+- [`ui.layout_columns()`](../layouts/arrange.qmd#grid-layouts)
+- `ui.div()`
+
+Here are some examples of non-container components:
+
+- [`ui.input_text()`](../components/inputs/text-box.qmd)
+- [`ui.output_plot()`](../components/outputs/plot-matplotlib.qmd)
+
+**In Shiny Core, to put a component inside of a container, you nest the function calls,** like `ui.sidebar(ui.input_text())`.
+
+::: {.column-page-right}
+```{shinylive-python}
+#| standalone: true
+#| components: [editor, viewer]
+#| layout: vertical
+# Core
+from shiny import ui, render, App
+
+app_ui = ui.page_sidebar(
+    ui.sidebar(
+        ui.input_text("txt_in", "Type something here:"),
+        open="always",
+    ),
+    ui.card(
+        ui.output_code("result"),
+    )
+)
+
+def server(input, output, session):
+    @render.code
+    def result():
+        return f"You entered '{input.txt_in()}'."
+
+app = App(app_ui, server)
+```
+:::
+
+By contrast, **in Shiny Express, container components like `ui.sidebar()` are _context managers_**, and used via `with` statements.
+Their child components go within the `with` block.
+
+::: {.column-page-right}
+```{shinylive-python}
+#| standalone: true
+#| components: [editor, viewer]
+#| layout: vertical
+# Express
+from shiny import render, App
+from shiny.express import input, ui
+
+with ui.sidebar():
+    ui.input_text("txt_in", "Type something here:")
+
+with ui.card():
+    @render.code
+    def result():
+        return f"You entered '{input.txt_in()}'."
+```
+:::
+
+##### Core advantages
+
+- Passing children as arguments is beautifully simple and robust from a programming perspective.
+- Containers and non-containers behave very similarly (in fact, non-containers act no different than containers with no children).
+- All of the Core UI components are simple value-returning functions, without side effects. This makes them easy to compose, reuse, refactor, inspect, and test.
+
+##### Express advantages
+
+- The `with` block syntax is slightly more readable, as it depends only on indentation and not on parentheses and commas.
```

Comment:
I don't love more readable since we say Core is more readable in other contexts. Maybe "More familiar to many users" or "easier to write because you don't have to worry about commas"?

Reply:
```suggestion
- The `with` block syntax is more forgiving to write, as you don't need to separate elements with commas. (Every Shiny Core app author has probably seen the "Perhaps you forgot a comma?" error message many times.)
```